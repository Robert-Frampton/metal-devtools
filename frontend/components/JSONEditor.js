import Component, {Config} from 'metal-jsx';
import JSONEditor from 'metal-json-editor';

import {keys, isEqual} from 'lodash';

class MetalJSONEditor extends Component {
  arrowRenderer(expanded) {
    return <span class={expanded ? 'arrow down' : 'arrow right'} />;
  }

  handleOnChange(newValue) {
    const {onChange, type, value} = this.props;

    const changedData = {};

    keys(value).forEach(key => {
      const keyVal = value[key];
      const newKeyValue = newValue[key];

      if (!isEqual(keyVal, newKeyValue)) {
        changedData[key] = newKeyValue;
      }
    });

    onChange(changedData, type);
  }

  render() {
    const {typeColors, value} = this.props;

    return (
      <JSONEditor
        arrowRenderer={this.arrowRenderer.bind(this)}
        data={value}
        onChange={this.handleOnChange.bind(this)}
        typeColors={typeColors}
      />
    );
  }
}

MetalJSONEditor.PROPS = {
  config: Config.object(),
  onChange: Config.func(),
  type: Config.string(),
  typeColors: Config.object(),
  value: Config.object()
};

export default MetalJSONEditor;
