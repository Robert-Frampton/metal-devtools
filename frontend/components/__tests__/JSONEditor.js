jest.unmock('../JSONEditor');

import jsoneditor from 'jsoneditor';

import JSONEditor from '../JSONEditor';

describe('JSONEditor', () => {
	test('should render', () => {
		const component = new JSONEditor();

		expect(component).toMatchSnapshot();
	});

	test('should return only the data that was changed', () => {
		const component = new JSONEditor({
			value: {
				baz: 'qux',
				foo: 'bar'
			}
		});

		component.getData = jest.fn(
			() => ({
				baz: 'qux',
				foo: 'boo'
			})
		);

		const retVal = component.getChangedData();

		expect(retVal).toEqual({foo: 'boo'});
	});

	test('should return the jsoneditor value', () => {
		const component = new JSONEditor();

		component.props.value = {foo: 'bar'};

		jest.runAllTimers();

		expect(component.getData()).toEqual({foo: 'bar'});
	});

	test('should return jsoneditor instance', () => {
		const component = new JSONEditor();

		expect(component.getEditor()).toBeInstanceOf(jsoneditor);
	});

	test('should call onChange prop', () => {
		const spy = jest.fn();

		const component = new JSONEditor({
			onChange: spy
		});

		component.handleChange();

		jest.runAllTimers();

		expect(spy).toHaveBeenCalled();
	});

	test('should not call render', () => {
			const component = new JSONEditor();

			component.render = jest.fn();

			component.props.value = {foo: 'bar'};

			expect(component.render).not.toHaveBeenCalled();
			expect(component.shouldUpdate()).toBe(false);
	});

	test('should set new jsoneditor value', () => {
		const component = new JSONEditor();

		component._editor.set = jest.fn();

		component.props.value = {foo: 'bar'};

		jest.runAllTimers();

		expect(component._editor.set).toHaveBeenCalled();
	});
});
