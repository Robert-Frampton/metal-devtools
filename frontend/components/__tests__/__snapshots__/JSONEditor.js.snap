exports[`JSONEditor should render 1`] = `
<div>
  <div class="jsoneditor jsoneditor-mode-form">
    <div class="jsoneditor-menu">
      <button type="button" class="jsoneditor-expand-all" title="Expand all fields"></button>
      <button type="button" title="Collapse all fields" class="jsoneditor-collapse-all"></button>
      <table class="jsoneditor-search">
        <tbody>
          <tr>
            <td>
              <div class="jsoneditor-results"></div>
            </td>
            <td>
              <div class="jsoneditor-frame" title="Search fields and values">
                <table>
                  <tbody>
                    <tr>
                      <td>
                        <button type="button" class="jsoneditor-refresh"></button>
                      </td>
                      <td>
                        <input>
                      </td>
                      <td>
                        <button type="button" title="Next result (Enter)" class="jsoneditor-next"></button>
                      </td>
                      <td>
                        <button type="button" title="Previous result (Shift+Enter)" class="jsoneditor-previous"></button>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="jsoneditor-outer">
      <div class="jsoneditor-tree">
        <table class="jsoneditor-tree">
          <colgroup>
            <col width="24px">
            <col>
          </colgroup>
          <tbody></tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
